machinelearning/horovod/README.md

```diff
@@ -0,0 +1,127 @@
+# Horovod
+
+Horovod is a distributed training framework for TensorFlow, Keras, and PyTorch. The goal of Horovod is to make distributed Deep Learning fast and easy to use.
+
+This is a port of the Dockerfile maintained at https://github.com/uber/horovod
+
+## Instructions
+
+A difference between the running the singularity container (vs the docker one) is that mpi runs outside of singularity thus you do not have to start the container instance on each host as detailed in https://github.com/uber/horovod/blob/master/docs/docker.md .
+
+### Build instructions
+Download the [horovod.def](https://github.com/sylabs/examples/tree/master/machinelearning/Horovod/horovod.def) file to your local computer.
+
+Build the sif file by executing the following command:
+
+```
+   sudo singularity build horovod.sif horovod.def
+```
+
+Copy the resultant `horovod.sif` file to all nodes you wish to run on (global storage can also be used).
+
+### Infrastructure Requirements
+Singularity 3.0 must be installed on all of the nodes you are running on.
+
+The resultant `horovod.sif` file (from the above build step) must be located in the same directory structure on every node.
+
+The same (or ABI compatible) version of OpenMPI has to be installed on all of the nodes you intend to run on (**NOTE:** this host version must match the OpenMPI version, or be ABI compatible, with the MPI in the container).
+
+Nvidia GPU's must be installed and the CUDA/Nvidia drivers must be installed on the nodes (at least one per node).
+
+An existing network infrastructure running across all nodes you intend to run on.
+
+The ability to SSH between all of the nodes (for the user running horovod) without having to supply passwords (ssh keys setup).
+
+Knowledge of MPI and mca parameters (these are specific to your site installation/network).
+
+For instance, in the example runs below, native IB was used for the network connections as well as IPoIB for sideband communications. The names of the nodes were f1, f2, f3 and f4. Also, f1:2 would specify to use node f1 and to use 2 processes on that node.
+
+### Running horovod using mpirun
+
+Run the horovod singularity container as you would run any other MPI job. You only have to start the execution on a single node, mpirun does the rest.
+
+**NOTE:** Resource schedulers such as SLURM can also be used to invoke the mpi execution.
+
+
+### Example runs
+
+Here is the included example `keras_mnist_advanced.py` run on 2 nodes with 2 processes per node. Each process with be using a GPU in the node:
+
```

Comment:
`Each process "will"`? correct me if I'm wrong @ctmadison

Reply:
No ... each process is playing *with* a GPU... Right? Right? _looks around_ ... Right? 

(Yeah... it should be will)